Issue #292 - Add config files for kunpeng2 (Linux ARM64)

martin-g opened the pull request:
kunpeng2 will be an external build node that will executed the builds and checks on Linux ARM64 (OS: openEuler 22.03 SP1 LTS).

Maintainers:
- Martin Grigorov (martin.grigorov@gmail.com)
- Yikun Jiang (yikunkero@gmail.com)

martin-g committed:
Issue #292 - Add config files for kunpeng2 (Linux ARM64)

kunpeng2 will be an external build node that will executed the builds
and checks on Linux ARM64 (OS: openEuler 22.03 SP1 LTS).

Maintainers:
- Martin Grigorov (martin.grigorov@gmail.com)
- Yikun Jiang (yikunkero@gmail.com)

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
Since we are going to maintain the external build node we would prefer to use openEuler Linux distro instead of Ubuntu. This is what the majority of our users use!
Please let us know if you think that won't be a good idea for any reason!

martin-g commented on 3.18/bioc/kunpeng2/config.sh:
```diff
@@ -0,0 +1,47 @@
+# Settings for kunpeng2 (Linux ARM64)
+# ===================================
+
+# Specific for openEuler START because some libraries are installed from source
```
Some OS dependencies are not available yet in openEuler and because of this they are installed from source. We work on adding RPM packages for them!

Currently these are the missing/mismatching ones:
```
from apt-bioc.txt
    libsbml5-dev - Installed from sources
    libopenbabel-dev - Installed from sources
    clustalo
    ocl-icd-opencl-dev
    libavfilter-dev
    kallisto
    mono-runtime
    libmono-system-data4.0-cil
    collectl
    gsl (Bioc requires gsl-devel 2.5+ while oe22.03 provides gsl-2.4-9.oe2203sp1.aarch64). Update: installed from             sources

from apt-cran.txt
    libsasl2-dev - could we use cyrus-sasl-devel instead ?!
    libcurl4-openssl-dev - installed libcurl-devel
    mpi-default-dev - using openmpi instead
    libudunits2-dev - Installed from sources
    libmysqlclient-dev - installed mysql-devel
    libgdal-dev
    jags - installed from sources

from apt_extra_fonts.txt
    gsfonts-x11
    xfonts-base
    xfonts-scalable
    xfonts-100dpi
    xfonts-75dpi
    t1-xfree86-nonfree
    ttf-xfree86-nonfree
    ttf-xfree86-nonfree-syriac

from apt_nice_to_have.txt
    manpages-dev
    from apt_optional_compile_R.txt
    gobjc
    libjpeg-dev
    libcairo2-dev
    libicu-devel

from apt_vignettes_reference_manuals.txt
    texlive-font-utils
    texlive-latex-extra
    texlive-fonts-extra
    texlive-bibtex-extra
    texlive-science
    texlive-lang-european
    pandoc - installed from sources
    pandoc-citeproc - installed from sources
    biber
    texlive-epic 
    texlive-xy 
    texlive-dsfont
```

Some of them are installed from their sources, others just have different package names in the RPM distros.

hpages replied:
Environment variables like this don't control BBS and are not related to it either. It would be much better to define them in a place that makes them available to all users on your machine e.g. in `/etc/profile` (that's where we define them on our Linux and Mac builders).

martin-g replied:
I already have them in the profile but I also put them here so that we have them in Git.
How about moving them to nodes/kunpeng2/local-settings.sh ?

hpages replied:
> How about moving them to nodes/kunpeng2/local-settings.sh ?

Yes, that's a better place to have them.

martin-g commented:
Would it be possible to add custom content to the generated report too ?
We think it would be good to have a short section `How to debug Linux ARM64 related issues on a x86_86 host`. The information is available in my blog - https://martin-grigorov.medium.com/emulated-building-and-testing-of-bioconductor-projects-for-linux-arm64-501c1f854dc1 but we want to make it easier to find to the package maintainers.

hpages commented:
> Would it be possible to add custom content to the generated report too ?

One way to go is that you add the following field to the current entry for kunpeng2 in `nodespecs.py`:
```
   'kunpeng2': {'OS'      : "Linux (openEuler 22.03 LTS-SP1)",
                'Arch'    : "aarch64",
                'Platform': "aarch64-linux-gnu",
                'pkgType' : "source",
                'encoding': "utf-8",
                'displayOnHTMLReport': "some-HTML-that-you-want-to-see-displayed-on-the-daily-report-for-kunpeng2-results"
          ''}
```
Then we'll make the necessary changes to the `BBS-report.py` code to inject the value of that field to the pages that display results for kunpeng2. For example, you could set `displayOnHTMLReport` to something like this:
```
                'displayOnHTMLReport': "See Martin Grigorov's <A href=\"https://martin-grigorov.medium.com/emulated-building-and-testing-of-bioconductor-projects-for-linux-arm64-501c1f854dc1\">blog post</A> for how to debug Linux ARM64 related issues on a x86_64 host."
```

Would that work?

martin-g committed:
Issue #292: Add displayOnHTMLReport setting for kunpeng2

Its value will be added to the index page of the report with a link to
an article explaining how to debug issues on Linux aarch64

Suggested-at: https://github.com/Bioconductor/BBS/pull/293#issuecomment-1577238357

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
> Would that work?

Sounds very good! I've made the change!

martin-g committed:
Issue #292: Move kunpeng2 specific env vars to its local-settings.sh

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

hpages commented:
Looks good on my end. Ready to merge?

martin-g commented:
Ready!

On Mon, 5 Jun 2023 at 22:39, Hervé Pagès ***@***.***> wrote:

> Looks good on my end. Ready to merge?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Bioconductor/BBS/pull/293#issuecomment-1577368677>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABYUQVPOQF77OAUA2PUDNLXJYYYNANCNFSM6AAAAAAY2Z7LSA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hpages commented:
Done. Do you think you can start `run.sh` tomorrow at 3pm EST?

martin-g commented:
> Do you think you can start `run.sh` tomorrow at 3pm EST?

Yes!

I'll also make `~/bbs-3.18-bioc/products-out/` available via HTTP! I'll send you the full address privately in Slack!

We added `155.52.207.166` (nebbiolo2) to our firewall rules. Please let us know if another IP will make the `rsync` call to fetch the output files!

martin-g commented:
Here is the last full (as a standalone runner) report by `kunpeng2` - https://bio-arm.github.io/openeuler-0606/report/long-report.html
- 27 failed INSTALLs
- 141 failed BUILDs
- 137 failed CHECKs

Today's nebbiolo2 [report](https://bioconductor.org/checkResults/3.18/bioc-LATEST/long-report.html):
- 10 failed INSTALLs
- 55 failed BUILDs
- 137 failed CHECKs

martin-g commented:
Here is the content of the scheduled cron job:
```
$ cat /etc/cron.d/bioconductor-BBS 
USER=biocbuild
SHELL=/bin/bash

# By default, PATH is set to /usr/bin:/bin only. We need /usr/local/bin
# for things that we install locally (e.g. new versions of pandoc).
# It must be placed **before** /usr/bin.
PATH=/home/biocbuild/bin:/usr/local/bin:/usr/bin:/bin

# BIOC 3.18 SOFTWARE BUILDS at 15h EST (19h GMT)
# -------------------------

# run:
00 19 * * 0-5 /bin/bash --login -c 'cd /home/biocbuild/BBS/3.18/bioc/`hostname` && ./run.sh >>/home/biocbuild/bbs-3.18-bioc/log/`hostname`-`date +\%Y\%m\%d`-run.log 2>&1'
```

hpages commented:
Looks good. Make sure the cron job will be run from the `biocbuild` account (I don't know if it's enough to set `USER=biocbuild` to guarantee that). Easiest way to do this is to set the cron job from the `biocbuild` account. Then you should be able to see `biocbuild`'s crontab with `crontab -l` (from the `biocbuild` account) or with `crontab -u biocbuild -l` (from any sudoer account).

Note that the last build report for kunpeng2 still has many timeouts. This could maybe come from the fact that you're overloading the machine by trying to parallelize the builds too much. I see that you are using the following settings on the machine:
```
export BBS_NB_CPU=28         # 32 cores are available
export BBS_BUILD_NB_CPU=20   # 32 cores are available
export BBS_CHECK_NB_CPU=25   # 32 cores are available
```
Maybe try to reduce `BBS_BUILD_NB_CPU` and `BBS_CHECK_NB_CPU` a little bit e.g. to 16 and 20, respectively, and see if that helps.

Thanks

martin-g commented:
The cron job started!

Actually I was using 14/14 until now - https://github.com/martin-g/BBS/pull/1/files#diff-7c5b3923098efc8afc99952be0ab603cfda21d29718aa5b73b9e0c4a47e2239eR26-R28
I increased them because I noticed that several CPUs stayed idle in the last two runs. Let's see how it goes and I will tune them!
The timeouts actually happen when downloading big data packages (1GB and bigger). I was able to overcome this problem by increasing `R_DEFAULT_INTERNET_TIMEOUT` - https://github.com/martin-g/BBS/pull/1/files#diff-9bde86ba458bcf9244b1761788e2fc146abe8d7ac66da7f402f29998ed8c437bR54 . The reason is that some packages download big files from some third party websites and the connection from Singapore to these sites is not that good.

hpages commented:
> The cron job started!

yeaah!!! :rocket: 

> The reason is that some packages download big files from some third party websites and the connection from Singapore to these sites is not that good.

I see, thanks for the explanation. This actually points to a weakness in these packages. At least they should cache those big files (e.g. by using **BiocFileCache**) so they download them only once.

Yikun commented:
I have added `155.52.47.146` to our firewall rules. Pls let me know need more configure.

hpages commented:
Thanks @Yikun. What command should we use on nebbiolo2 (155.52.47.146) to retrieve the content of `~biocbuild/bbs-3.18-bioc/products-out/` from kunpeng2? It would be great if we could just use `rsync` for this. You can DM me the command on the community-bioc Slack if you don't want to share it here. Thanks!

Yikun commented:
I just realized we might also need also open some port for `rsync` and do some setup for `rsync server`, we will have a test today, and send the command to you.

hpages commented:
Sounds good. Just let me know when this is ready.
